tutorials/ImplementASolver.qmd

```diff
@@ -0,0 +1,340 @@
+
+- `initialize_solver!(mp, rws)` which initialises the solver before the first iteration
+- `step_solver!(mp, rws, i)` implement the `i`th iteration, where `i` is given to you as the third parameter
+- `get_iterate(rws)` to access the iterate from other places in the solver
+- `get_solver_result(rws)` toi return the final point we reached. Bz default this would
+  return the last iterate, but since our result is `q`, we have to set this
+
+The first two functions are in-place functions, that is they modify our solver state `rws`.
+You implement these by multiple dispatch on the types after importing said functions from Mantop:
+
+```{julia}
+import Manopt: initialize_solver!, step_solver!, get_iterate, get_solver_result
+```
+
+The state above has two fields where we use the common names used in `Manopt.jl`,
+that is the [`StoppingCriterion`](@ref) is usually in `stop` and the iterate in `p`.
+If your choice is different, you need to reimplement
+
+- `stop_solver!(mp, rws, i)` to determine whether or not to stop after the `i`th iteration.
+- `get_iterate(rws)` to access the current iterate
+
+We recommend to follow the general scheme with the `stop` field. If you have specific criteria
+when to stop, consider implementing your own [stoping criterion](https://manoptjl.org/stable/plans/stopping_criteria/) instead
+
+### Initialization & Iterate Access
+
+For our solver, there is not so much to initialize, just to be safe we should copy over the
+initial value in `p` we start with, to `q`. We do not have to care about remembering the iterate,
+that is done by `Manopt.jl`. For the iterate access we just have to pass `p`.
+
+```{julia}
+#| output: false
+function initialize_solver!(mp::AbstractManoptProblem, rws::RandomWalkState)
+    copyto!(M, rws.q, rws.p) # Set p^{(0)} = q
+    return rws
+end
+get_iterate(rws::RandomWalkState) = rws.p
+get_solver_result(rws::RandomWalkState) = rws.q
+```
+
+and siilarly we implement the step. Here we make use of the fact that the problem
+(and also the objective in fact) have access functions for their elements,
+the one we need is [`get_cost`](@ref).
+
+```{julia}
+#| output: false
+function step_solver!(mp::AbstractManoptProblem, rws::RandomWalkState, i)
+    M = get_manifold(mp) # for ease of use get the manifold from the problem
+    X = rand(M; vector_at=p)     # generate a direction
+    X .*= rws.σ/norm(M, p, X)
+    # Walk
+    retract!(M, rws.p, rws.p, X, rws.retraction_method)
+    # is the new point better? Then store it
+    if get_cost(mp, rws.p) < get_cost(mp, rws.q)
+        copyto!(M, rws.p, rws.q)
+    end
+    return rws
+end
+```
+
+Performance wise we could improve the number of allocations by making `X` also a field of
+our `rws` but let's keep it simple here.
+We could also store the cost of `q` in the state, but we will see how to easily also enable
+this solver to allow for [caching](https://manoptjl.org/stable/tutorials/CountAndCache/#How-to-Count-and-Cache-Function-Calls).
```

Comment:
```suggestion
this solver to allow for [caching](https://manoptjl.org/stable/tutorials/CountAndCache/#How-to-Count-and-Cache-Function-Calls).
In practice, however, it is preferable to cache intermediate values like cost of `q` in the state when it can be easily achieved. This way we do not have to deal with overheads of an external cache.
```